tools/cosmovisor/cmd/cosmovisor/batch_upgrade.go

```diff
@@ -0,0 +1,138 @@
+	cfg, err := getConfigFromCmd(cmd)
+	if err != nil {
+		return err
+	}
+	upgradeFile, err := cmd.Flags().GetString("upgrade-file")
+	if err == nil && upgradeFile != "" {
+		info, err := os.Stat(upgradeFile)
+		if err != nil {
+			return fmt.Errorf("error getting reading upgrade file %s: %w", upgradeFile, err)
+		}
+		if info.IsDir() {
+			return fmt.Errorf("upgrade file %s is a directory", upgradeFile)
+		}
+		return processUpgradeFile(cfg, upgradeFile)
+	}
+	upgradeList, err := cmd.Flags().GetStringSlice("upgrade-list")
+	if err != nil || len(upgradeList) == 0 {
+		return fmt.Errorf("either --upgrade-file or --upgrade-list must be provided")
+	}
+	return processUpgradeList(cfg, upgradeList)
+}
+
+// processUpgradeList takes in a list of upgrades and creates a batch upgrade file
+func processUpgradeList(cfg *cosmovisor.Config, upgradeList []string) error {
+	upgradeInfoPaths := []string{}
+	for i, as := range upgradeList {
+		a := strings.Split(as, ":")
+		if len(a) != 3 {
+			return fmt.Errorf("argument at position %d (%s) is invalid", i, as)
+		}
+		upgradeName := filepath.Base(a[0])
+		upgradePath := a[1]
+		upgradeHeight, err := strconv.ParseInt(a[2], 10, 64)
+		if err != nil {
+			return fmt.Errorf("upgrade height at position %d (%s) is invalid", i, a[2])
+		}
+		upgradeInfoPath := cfg.UpgradeInfoFilePath() + "." + upgradeName
+		upgradeInfoPaths = append(upgradeInfoPaths, upgradeInfoPath)
+		if err := addUpgrade(cfg, true, upgradeHeight, upgradeName, upgradePath, upgradeInfoPath); err != nil {
+			return err
+		}
+	}
+
+	var allData []json.RawMessage
+	for _, uip := range upgradeInfoPaths {
+		fileData, err := os.ReadFile(uip)
+		if err != nil {
+			return fmt.Errorf("error reading file %s: %w", uip, err)
+		}
+
+		// Verify it's valid JSON
+		var jsonData json.RawMessage
+		if err := json.Unmarshal(fileData, &jsonData); err != nil {
+			return fmt.Errorf("error parsing JSON from file %s: %w", uip, err)
+		}
+
+		// Add to our slice
+		allData = append(allData, jsonData)
+	}
+
+	// Marshal the combined data
+	batchData, err := json.MarshalIndent(allData, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshaling combined JSON: %w", err)
+	}
+
+	// Write to output file
+	err = os.WriteFile(cfg.UpgradeInfoBatchFilePath(), batchData, 0o600)
+	if err != nil {
+		return fmt.Errorf("error writing combined JSON to file: %w", err)
+	}
+
+	return nil
+}
+
+// processUpgradeFile takes in a batch upgrade file, validates it and copies it to the upgrade directory
+func processUpgradeFile(cfg *cosmovisor.Config, upgradeFile string) error {
+	b, err := os.ReadFile(upgradeFile)
+	if err != nil {
+		return fmt.Errorf("error reading upgrade file %s: %w", upgradeFile, err)
+	}
+	var batch []upgradetypes.Plan
+	if err := json.Unmarshal(b, &batch); err != nil {
+		return fmt.Errorf("error unmarshalling upgrade file %s: %w", upgradeFile, err)
+	}
+	return copyFile(upgradeFile, cfg.UpgradeInfoBatchFilePath())
```

Comment:
this looks incomplete. Where is batch used? And why aren't we calling add upgrade there as well?

Reply:
This is no different than `processUpgradeList` where we end with writing out the batch file to the home directory (https://github.com/cosmos/cosmos-sdk/blob/1d21f555bdf0e24520c1d95116a1febc2eca619f/tools/cosmovisor/cmd/cosmovisor/batch_upgrade.go#L119). Here we're already getting a batch file from the user so we just validate it and copy it over. We then watch this file during `cosmovisor run`.

Reply:
`processUpgradeList` calls `addUpgrade`, which place the binaries at the right place.
This doesn't.

Reply:
You're right, my bad. `upgradetypes.Plan` doesn't really have field we can use for the binary path, what do you think of changing the file format for `--upgrade-file` to be a CSV with rows `name,path/to/binary,height`? I'd imagine it'd be easier for users to make than a JSON file anyway.